kroxylicious/src/main/java/io/kroxylicious/proxy/internal/net/EndpointRegistry.java

```diff
@@ -0,0 +1,417 @@
+     *
+     * @param virtualCluster virtual cluster to be deregistered.
+     * @return completion stage that will complete after registration is finished.
+     */
+    public CompletionStage<Void> deregisterVirtualCluster(VirtualCluster virtualCluster) {
+        Objects.requireNonNull(virtualCluster, "virtualCluster cannot be null");
+
+        var deregisterFuture = new CompletableFuture<Void>();
+        var vcr = registeredVirtualClusters.get(virtualCluster);
+        if (vcr == null) {
+            // cluster not currently registered
+            deregisterFuture.complete(null);
+            return deregisterFuture;
+        }
+
+        var updated = vcr.deregistrationStage().compareAndSet(null, deregisterFuture);
+        if (!updated) {
+            // cluster de-registration already in progress.
+            return vcr.deregistrationStage().get();
+        }
+
+        var unused = vcr.registrationStage()
+                .thenCompose((u) -> deregisterBinding(virtualCluster, binding -> binding.virtualCluster().equals(virtualCluster))
+                        .handle((unused1, t) -> {
+                            registeredVirtualClusters.remove(virtualCluster);
+                            if (t != null) {
+                                deregisterFuture.completeExceptionally(t);
+                            }
+                            else {
+                                deregisterFuture.complete(null);
+                            }
+                            return null;
+                        }));
+        return deregisterFuture;
+    }
+
+    /* test */ boolean isRegistered(VirtualCluster virtualCluster) {
+        return registeredVirtualClusters.containsKey(virtualCluster);
+    }
+
+    /* test */ int listeningChannelCount() {
+        return listeningChannels.size();
+    }
+
+    private CompletionStage<Endpoint> registerBinding(Endpoint key, String host, VirtualClusterBinding virtualClusterBinding) {
+        Objects.requireNonNull(key, "key cannot be null");
+        Objects.requireNonNull(virtualClusterBinding, "virtualClusterBinding cannot be null");
+        var virtualCluster = virtualClusterBinding.virtualCluster();
+
+        var lcr = listeningChannels.computeIfAbsent(key, k -> {
+            // the listening channel doesn't exist, atomically create a record to represent it and request its binding to the network.
+            var future = new CompletableFuture<Channel>();
+            var r = ListeningChannelRecord.create(future.exceptionally(t -> {
+                // Handles the case where the network bind fails
+                listeningChannels.remove(key);
+                if (t instanceof RuntimeException re) {
+                    throw re;
+                }
+                else {
+                    throw new RuntimeException(t);
+                }
+            }));
+
+            queue.add(new NetworkBindRequest(key.bindingAddress(), key.port(), virtualCluster.isUseTls(), future));
```

Comment:
Worth constructing an Endpoint here? It could be nice to see the Endpoint term when we are passing these 3 things.